### ISSUE TITLE
simplify does not simplify complex numbers composed into real and imaginary part

### ISSUE THREAD

**Issue by laolux** [sympy/sympy#18271]
Currently simplify does not simplify complex numbers decomposed into real and imaginary part.
A simple example is to take a a complex number and subtract its real and imaginary part (*i). This should simplify to zero.

    import sympy
  
    z = sympy.symbols('z', complex=True)
    simple = sympy.simplify(z-(sympy.re(z)+sympy.I*sympy.im(z)))
    sympy.pprint(simple)
    simple2 = sympy.simplify(sympy.conjugate(z)-(sympy.re(z)-sympy.I*sympy.im(z)))
    sympy.pprint(simple2)

Another very cool feature for simplify would be to rewrite `sympy.re(z)+sympy.I*sympy.im(z)` to `z` whenever encountered in any expression.

**Comment by asmeurer** [comment]
Related to https://github.com/sympy/sympy/issues/10822. I remember this also being in another issue, but I can't find it right now.

**Comment by gschintgen** [comment]
There's also #7569.
```
In [23]: a = z-(re(z)+I*im(z))

In [24]: expand_complex(a)
Out[24]: 0
```

**Comment by harsh9200** [comment]
can i work on this ? any suggestions ?

**Comment by gschintgen** [comment]
> can i work on this ? any suggestions ?

Yes, sure. I think #7569 is a good start. You'll need to familiarize yourself with the basic working of `simplify` and decide where and how best to add the call to `expand_complex`. (Don't forget to run the test suite locally, this could have unanticipated side effects in other places.)

You could also try a more direct approach (e.g. reconstituting z from re(z) and im(z)).

**Comment by AaronStiff** [comment]
If this isn't currently being worked on, could I suggest the following?

```diff
diff --git a/sympy/core/add.py b/sympy/core/add.py
index ca9bcfc4a3..3e7e9012f1 100644
--- a/sympy/core/add.py
+++ b/sympy/core/add.py
@@ -279,11 +279,30 @@ def flatten(cls, seq):
             # s is an expression with number factor extracted
             # let's collect terms with the same s, so e.g.
             # 2*x**2 + 3*x**2  ->  5*x**2
+            from sympy.functions.elementary.complexes import re, im
+            from sympy import I, Wild
+
             if s in terms:
                 terms[s] += c
                 if terms[s] is S.NaN and not extra:
                     # we know for sure the result will be nan
                     return [S.NaN], [], None
+            # combine real and imaginary parts of complex number (issue #18271)
+            elif s.match(I*im(Wild('z'))) or s.match(re(Wild('z'))):
+                if isinstance(s, re):
+                    var = s.args[0]
+                    other_part = I*im(var)
+                else:
+                    var = s.as_ordered_factors()[1].args[0]
+                    other_part = re(var)
+                if other_part not in terms:
+                    terms[s] = c
+                elif not terms.get(var, False):
+                    terms[var] = c
+                    terms.pop(other_part)
+                else:
+                    terms[var] += c
+                    terms.pop(other_part)
             else:
                 terms[s] = c
```

I'd welcome suggestions/critiques (it feels a little messy to me). It solves the original issue, as well as the one mentioned which relates to `expand_complex`. I've run all tests in `core` and they pass.

**Comment by oscarbenjamin** [comment]
If I understand correctly that change is to `Add.flatten`. We have to be very careful making changes in `Add.flatten` because it is used everywhere throughout the codebase and it also has a big impact on the speed of sympy. Some of the functions you are using here look quite slow e.g. `match`.

Making a change in `Add.flatten` means that it happens automatically and automatic evaluation is actually quite problematic. The difficulty is that when we have automatic evaluation it becomes hard to work with the unevaluated expression. That in turn makes it difficult to convert expressions between different forms.

It would be better to do this as part of simplify. There could be a simplification function specifically for handling this kind of case. It could be called somewhere inside simplify.

**Comment by AaronStiff** [comment]
Yes, that is correct. 

That makes sense. I'll look into it.

**Comment by asmeurer** [comment]
It would be better to have this in a simplification function. I don't know if a good one already exists to put it in. Maybe something like `simplify_complex` which could do this case and also simplify similar instances of `abs`, `im`, `re`, `arg`, and `conjugate`.

**Comment by olcc** [comment]
Can you please also ensure that such expression gets simplified? Currently it does not.
```
var('q', zero=False)
simplify((re(q)**2+im(q)**2)/abs(q))
```

**Comment by mammalwong** [comment]
This issue is still exists in version 1.12, and it fail to combine the real term and imaginary term in this simple case:
```
import sympy as sy
x = sy.symbols('x')
(sy.re(x) + sy.I*sy.im(x)).simplify()
```
which output:
```
re(x)+i im(x)
```
